app/models/ont/run.rb

```diff
@@ -76,30 +87,72 @@ def cancel
       update(deactivated_at: DateTime.current)
     end
 
+    # Sets flowcells from an array of attributes (hash)
+    # If there is no id in attributes, a new flowcell will be created.
+    # If there is an id, existing flowcell will be updated.
+    # If there is no attributes hash for an existing flowcell, it will be
+    # deleted.
+    # This method is called before validations, therefore it must properly
+    # set up all flowcells to be saved together with the run. It must first
+    # do the deletions and then "build" flowcells and "assign" attributes
+    # without saving.
+
     def flowcell_attributes=(flowcell_options)
-      self.flowcells = flowcell_options.map do |attributes|
+      transform_flowcell_attributes(flowcell_options)
+
+      # Delete flowcells if attributes are not given
+      options_ids = flowcell_options.pluck(:id).compact
+      flowcells.each do |flowcell|
+        flowcells.delete(flowcell) if options_ids.exclude? flowcell.id
+      end
+
+      # Update existing flowcells or build new ones
+      flowcell_options.map do |attributes|
         if attributes[:id]
           update_flowcell(attributes)
         else
-          Ont::Flowcell.new(attributes)
+          create_flowcell(attributes)
         end
       end
     end
 
     private
 
+    # Change attributes before using them
+    # Strip and upcase flowcell_ids to accept case-insensitive barcodes.
+    def transform_flowcell_attributes(flowcell_options)
+      flowcell_options.each do |attributes|
+        # because it is a frozen string
+        if attributes[:flowcell_id]
+          attributes[:flowcell_id] = attributes[:flowcell_id]&.strip&.upcase
+        end
+      end
+    end
+
+    # Assing attributes to an existing flowcell. The flowcell is found by the
+    # id given in the attributes. If there is no flowcell with that id, an
+    # exception is raised.
     def update_flowcell(attributes)
-      id = attributes[:id].to_s
-      indexed_flowcells.fetch(id) { missing_flowcell(id) }
-                       .tap { |flowcell| flowcell.update(attributes) }
+      matching = flowcells.select { |flowcell| flowcell.id == attributes[:id] }
+      record_not_found(attributes[:id]) unless matching
+      assign_flowcell_attributes(matching[0], attributes)
     end
 
-    def missing_flowcell(id)
+    def record_not_found(id)
       raise ActiveRecord::RecordNotFound, "Ont flowcell #{id} does not exist"
     end
 
-    def indexed_flowcells
-      @indexed_flowcells ||= flowcells.index_by { |flowcell| flowcell.id.to_s }
+    # Create a new flowcell instance and assign attributes.
+    def create_flowcell(attributes)
+      new_flowcell = flowcells.build
+      assign_flowcell_attributes(new_flowcell, attributes)
+    end
+
+    def assign_flowcell_attributes(flowcell, attributes)
```

Comment:
The original implementation which was direct assignment to flowcells association inside flowcell_attributes= was failing because it did not consider deleting the existing ones first and did saving or updating of individual flowcells.

The new implementation deletes flowcells missing from the attributes, builds new ones, assign attributes. The run an and flowcells go to validation together and then saved. 

As for calling asking_flowcell_attributes for new and existing flowcells:
The method enables a uniform handling of flowcells and attributes
The method assigns only the attributes that flowcell can respond to (compared to mass-assignment or cherry-picking which attributes to assign.)